File: src/main/kotlin/onl/tesseract/srp/skill/domain/model/station/StationStats.java
package onl.tesseract.srp.skill.domain.model.station;

import onl.tesseract.srp.domain.skill.station.StatType;
import onl.tesseract.srp.skill.domain.model.bonus.Bonus;
import onl.tesseract.srp.skill.domain.model.bonus.StationBonus;
import org.jetbrains.annotations.NotNull;

public record StationStats(
        UpgradeLevel tier,
        UpgradeLevel quality,
        UpgradeLevel success,
        UpgradeLevel timeReduction
) {
    public StationStats() {
        this(new UpgradeLevel(1), new UpgradeLevel(0), new UpgradeLevel(0), new UpgradeLevel(0));
    }

    public int getStatLevel(@NotNull StatType statType) {
        return switch (statType) {
            case TIER -> tier.value();
            case QUALITY -> quality.value();
            case SUCCESS -> success.value();
            case TIME_REDUCTION -> timeReduction.value();
            default -> throw new IllegalStateException("Unexpected value: " + statType);
        };
    }

    public StationBonus getBonus() {
        return new StationBonus(
                new Bonus(quality().value()*0.02),
                new Bonus(success().value()*0.015),
                new Bonus(timeReduction().value()*0.01)
        );
    }

    public StationStats upgradeStat(@NotNull StatType statType) {
        return switch (statType) {
            case TIER -> new StationStats(new UpgradeLevel(tier.value() + 1), quality, success, timeReduction);
            case QUALITY -> new StationStats(tier, new UpgradeLevel(quality.value() + 1), success, timeReduction);
            case SUCCESS -> new StationStats(tier, quality, new UpgradeLevel(success.value() + 1), timeReduction);
            case TIME_REDUCTION -> new StationStats(tier, quality, success, new UpgradeLevel(timeReduction.value() + 1));
            default -> throw new IllegalStateException("Unexpected value: " + statType);
        };
    }
}

File: src/main/kotlin/onl/tesseract/srp/domain/skill/station/StationStats.kt
package onl.tesseract.srp.domain.skill.station

data class StationStats(
    // Stat levels (0 = no upgrade)
    val tierLevel: Int = 0,
    val qualityBonusLevel: Int = 0,
    val successBonusLevel: Int = 0,
    val timeReducLevel: Int = 0,
)
{

    // Quality bonus percentage (0.0 to 1.0)
    val qualityBonusPower: Double
        get() = calculatePower(qualityBonusLevel, 0.02)

    // Success bonus percentage (0.0 to 1.0)
    val successBonusPower: Double
        get() = calculatePower(successBonusLevel, 0.015)

    // Time reduction percentage (0.0 to 1.0)
    val timeReductionPower: Double
        get() = calculatePower(timeReducLevel, 0.01)

    fun getStatPower(statType: StatType): Double {
        return when (statType) {
            StatType.TIER -> tierLevel.toDouble()
            StatType.QUALITY -> qualityBonusPower
            StatType.SUCCESS -> successBonusPower
            StatType.TIME_REDUCTION -> timeReductionPower
            else -> 0.0
        }
    }

    /**
     * Calculates power value based on level using logarithmic scaling.
     * Power increases more slowly at higher levels to prevent overpowering.
     *
     * @param level Current level (0 = no upgrade)
     * @param baseIncrement Base increment per level (diminishing returns applied)
     * @return Power value between 0.0 and 1.0
     */
    private fun calculatePower(level: Int, baseIncrement: Double): Double {
        if (level == 0) return 0.0
        // Logarithmic scaling: level * baseIncrement with diminishing returns
        return minOf(1.0, level * baseIncrement / (1 + (level * 0.1)))
    }

    fun upgradeStat(statType: StatType): StationStats {
        return when (statType) {
            StatType.TIER -> this.copy(tierLevel = tierLevel + 1)
            StatType.QUALITY -> this.copy(qualityBonusLevel = qualityBonusLevel + 1)
            StatType.SUCCESS -> this.copy(successBonusLevel = successBonusLevel + 1)
            StatType.TIME_REDUCTION -> this.copy(timeReducLevel = timeReducLevel + 1)
            else -> this
        }
    }

    fun getStatLevel(statType: onl.tesseract.srp.domain.skill.station.StatType): Int {
        return when (statType) {
            StatType.TIER -> tierLevel
            StatType.QUALITY -> qualityBonusLevel
            StatType.SUCCESS -> successBonusLevel
            StatType.TIME_REDUCTION -> timeReducLevel
            else -> 0
        }
    }
}